pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,885 @@
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	go c.processWaitingCaptures()
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processWaitingCaptures() {
+	for {
+		capture := func() string {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			for c.numCaptures >= maxConcurrentCaptures || c.waitingCaptures == 0 {
+				c.cond.Wait()
+			}
+			c.numCaptures += 1
+			c.waitingCaptures -= 1
+			capture := <-c.waitingCapturesCh
+			c.startPacketCapture(capture)
+			c.captures[capture] = packetCapturePhaseRunning
+			return capture
+		}()
+		c.queue.Add(capture)
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		if !apierrors.IsNotFound(err) {
+			c.queue.AddRateLimited(key)
+			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+		} else {
+			c.queue.Forget(key)
+			delete(c.captures, key)
+		}
+	}
+	return true
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted the captured pcap file successfully", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete the captured pcap file", "name", pcName, "path", path)
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("the packet file %s already exists. This may be caused by an unexpected termination", filePath)
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
```

Comment:
maybe add a comment to explain that at least one of `pc.Spec.Source.Pod` or `pc.Spec.Destination.Pod` must be set, as that is guaranteed by the OpenAPI schema for the CRD.